package service;

message QtObject
{
	message Property
	{
		required string name = 1;
		required bool isWritable = 2;

		// value of the property serialized from a QVariant
		// using QDataStream
		optional bytes value = 3;
	}

	// basic object properties
	required int32 id = 1;
	required string className = 2;
	optional string objectName = 3;
	repeated int32 childId = 5;
	optional int64 address = 6;
	
	// type-specific object properties
	repeated Property property = 4;
}

message InspectorRequest 
{
	enum RequestType
	{
		FetchObjectTree = 1;
		FetchObjectRequest = 2;
		WritePropertyRequest = 3;
		PickWidgetRequest = 4;
	}

	required RequestType type = 1;

	optional int32 objectId = 2;

	// for object fetch requests, indicates whether the properties
	// of objects should be fetched or not.  If false, only the basic
	// attributes of each object (id, className, objectName, children)
	// are fetched
	optional bool fetchProperties = 4;

	optional QtObject.Property propertyUpdate = 3;
}

message InspectorResponse
{
	required InspectorRequest.RequestType request = 1;

	// for FetchObjectTree requests, this contains the IDs
	// of the root objects
	repeated int32 rootObjectId = 3;
	repeated QtObject object = 2;
}

